### ISSUE TITLE
Multiple channels per delivery type

### ISSUE THREAD

**Issue by aienabled** [RevenantX/LiteNetLib#67]
Hello!
Are there any way of creating more than one sequenced channel? 
In Lidgren ( https://github.com/lidgren/lidgren-network-gen3/wiki/Sequence-Channels  ) it's possible to have up to 31 sequence channels per any sequenced delivery method.
I looked into the LiteNetLib source and it seems all the channels are created only once in NetPeer constructor.
For me, it makes "sequenced" channel almost useless - I cannot even send health in one channel and ammo in another channel. Only the latest received NetPacket will be used - either ammo or health.

Regards!

**Comment by RevenantX** [comment]
Hi! There is only one channel per message type. I can implement if you really need. 
But library have option "MergeEnabled" that will merge packets (health and ammo) together if you send them immediately.
You can send ammo and health together by hands (this is really small overhead. UDP header is 68 bytes).
You can use Unreliable with manual sequence and "channels" :)
I think multiple channels is rarely needed.
But if you have opposite opinion - i will listen)

**Comment by aienabled** [comment]
I see. Well, it's not a satisfying for my needs.
I mentioned health and ammo as an example. In most games there are much more things which need to be send in sequenced manner but in dedicated sequenced. It's simply impossible (or too much extra network traffic) to put all the data into a single packet just because we cannot have dedicated sequences.

In my case, the idea is to have multiple channels per game entity for replication needs. The channels should be dynamically creatable and disposable when not needed anymore.
For example, we need to replicate a few dozens of actors. Each one has its own properties (for example stats - such as health, shield, energy). Every property change needs to be replicated in a dedicated sequenced channel - dedicated to a specific actor. So each actor will have a few dedicated sequenced channels and network layer will ensure that the client has the fresh properties of a particular actor and no data is lost in transmission.

Of course, it's possible to a custom "data" layer over the networking library (as mostly done in AAA games with custom engines). So this "data" layer will send the data in a single unreliable channel and manually do the reliablity and sequencing procedures to ensure all the data is transferred as expected. But it's a lot of work to do and it will have overhead.

Much better to have this functionaly in the network library (as done in Lidgren).
The only problem with Lidgren is that it cannot dynamically allocate/dispose channels and the channels amount is limited to 31 per (sequenced) delivery method. I have hacked all that into my Lidgren fork but don't really like the code and now consider other networking libraries.

**Comment by RevenantX** [comment]
@aienabled do you need multiple channels only with "Sequenced" method?

**Comment by aienabled** [comment]
@RevenantX, no, in my game I already use this with any net delivery method which uses sequencing - such as unreliable sequenced, reliable unordered, reliable sequenced, reliable ordered.

**Comment by RevenantX** [comment]
@aienabled in what situation you need multiple channels with "reliable unordered" message type?

**Comment by aienabled** [comment]
@RevenantX, if player need to receive data as soon as possible and packet order is not important. For example, chat incoming messages. The ordering could be applied later based on the timestamp of the message. But it's important to receive all incoming chat messages.
Thought, you're right, it's very rarely used and mostly could be replaced with reliable ordered without much problem to the player.

**Comment by aienabled** [comment]
The thing is, in Lidgren network you could specify sequence channel for any outgoing message. It's not used for the unreliable delivery method, but for any other delivery method, there are 31 channels.

**Comment by RevenantX** [comment]
@aienabled I understand useful scenario of multiple channels with "Sequenced" type and "Reliable Ordered" :) But not with "reliable unordered"

**Comment by aienabled** [comment]
@RevenantX, I think, if you can make it generalized for other delivery methods, it will be no problem to have it for "reliable unordered" also. That's how it done in Lidgren.

**Comment by RevenantX** [comment]
@aienabled I can implement this feature only for "Sequenced" and "Reliable Ordered". Because this feature will add little overhead in packet size, used memory and performance. And maybe this feature will be optional.

**Comment by aienabled** [comment]
@RevenantX, that's great! Especially if I will be able to dynamically allocate channels and dispose them when they're not needed anymore (i.e. packets are acknowledged and channel is empty/drained). Up to 256 channels per delivery method will be enough for our needs.
Regarding "optional feature" - of course, not everyone need channel numbers (and extra byte per packet to determine channel number) for this. It could be optional with a compile flag or runtime boolean flag, I have nothing against it.

BTW, I missed - there are no unreliable sequenced channels? That's quite useful for me, for example, to send input - I never need to re-send it (no reliability is required as fresh data will be sent soon, it's simply sent with some interval) but it's important to use only latest input. Of course, I could add input Id number to determine if we have newer data, but it's simply better to have such option in network library.

**Comment by RevenantX** [comment]
@aienabled 
>BTW, I missed - there are no unreliable sequenced channels?

Sequenced - is unreliable sequenced. 
There is no "Lidgren" method "reliable sequenced".

**Comment by aienabled** [comment]
@RevenantX, wow, that's frustrating. I just checked my code for ReliableSequenced usage and found that I'm using it already in 21 places and where it's important to reduce traffic (by sequencing), but have reliablity. It's most popular network delivery method for me. For example, character stats (health/energy/etc), crafting progress, weapon firing mode. It's all important to have, reliably, latest value, but with reduced trafic (by sequencing - not re-sending late data). Should I create another issue here for this feature?

**Comment by RevenantX** [comment]
> Should I create another issue here for this feature?
If you need this feature - sure. :)

**Comment by aienabled** [comment]
Thanks. Reported here https://github.com/RevenantX/LiteNetLib/issues/68

**Comment by RevenantX** [comment]
@miketheprogrammer i will add multichannel mode after i complete `ReliableSequenced` packet type)

**Comment by aienabled** [comment]
I see there is still no multichannel support, though a lot of great improvements are done and I'm really impressed with all these changes and overall project activity.

I wish to try LiteNetLib with my MMO project I talked about before here ([game page](http://cryofall.com), now available in a free open alpha and getting very positive reviews; currently using `Lidgren.Network`) but unfortunately multiple channels is one of the requirements. It's possible but not reasonable to do on the application level (by using unreliable channel and manually sending and processing ACKs) - I believe it will be beneficial to have this feature for everyone, implemented and perfected in the network library itself.

Ability to transfer certain packets as sequenced (with or without reliability) is very useful. But the usefulness of the sequenced channel is very limited if there is only a single channel of this kind for the whole game connection - it could be beneficial for a certain kind of session-based games like a FPS shooter where a basic set of data is replicated every other frame and client should receive only the latest version of this data (but in that case it's actually could be easily implemented on an application level just by checking the relative sequence number of the packet). For any bigger project (like an MMO game I'm building) multiple channels support is a real necessity and implementing this in a proper way (performance-wise) is a serious challenge.

Also, I would prefer to contribute in an open-source project to make this feature useful and optimized for everyone than to implement it on an application level specifically for my game engine.

Regards!

**Comment by RevenantX** [comment]
@aienabled i am working on this feature. Just want to make it with maximum usability, so it takes some time)

**Comment by TimonPost** [comment]
I've been researching this topic lately. I saw the same topic in Raknet. RakNet is a C++ lib like this one. They call it ordering streams. You could use this as a reference: http://www.raknet.net/raknet/manual/sendingpackets.html

**Comment by rikimaru0345** [comment]
@RevenantX 
I assume this will be implemented (correct me if I'm wrong) and I have a question.

How will the implementation look like roughly?
Will we be able to set what channels there are and what their transport settings are?

Like before/while constructing the NetManager maybe we could do something like:

`AddChannel(id: 1, deliveryMethod: DeliveryMethod.ReliableOrdered, debugChannelName: "abc");`

and LiteNetLib would just ensure that the channel IDs start at 1, there are no IDs skipped, no IDs are assigned twice... and of course some upper limit (32 channels at most I thought was mentioned at some point?)

**Comment by RevenantX** [comment]
@rikimaru0345 
>I assume this will be implemented (correct me if I'm wrong) and I have a question.

Yes. There is work in progress.
Right now there is some things that i don't solved, so i cannot say anything this)

**Comment by Krakean** [comment]
@RevenantX any progress on this ticket?

**Comment by RevenantX** [comment]
@Krakean https://github.com/RevenantX/LiteNetLib/tree/channels

**Comment by RevenantX** [comment]
https://github.com/RevenantX/LiteNetLib/commit/8d7336e68c57422189161f94fea60f0c2d87b3c4 integrated to master
